Ci 1008 seo add a breadcrumb list to article json ld metadata

juanSanchezAlcala opened the pull request:
Why ?
There is a recommendation to improve SEO that we should set a breadcrumb ld+json script into every article. see [ticket](https://financialtimes.atlassian.net/browse/CI-1008)
How?
Since we don't have a breadcrumb we are going to build or breadcrumb for SEO in the next way:

-We will search in annotations all items with "implicitAbout" and "about" items. The order of preference is "implicitAbout" go first, in case a conflict we will pick those who have less slashes in their relativeUrl , in case of conflict with predicate and slashes we will pick in alphabetical order the prefLabel. The prefLabel will be the name used to build the breadcrumb. 
-If the item has a displayConcept property this annotation will be set as the last item of the breadcrumb
-The breadcrumb will have a maximum of 3 items

Using this approach the breadcrumb will remain the same even if the order of annotations change trying to have sense in order of the information that we can gather from annotations.

I attach a report in json and csv format with info of the articles and their breadcrumb ld+json. This report was generated with "puppeteer" navigating throw pages, so it could be reply easily in any change.
[log.csv](https://github.com/Financial-Times/next-json-ld/files/8621749/log.csv)
[log.json.txt](https://github.com/Financial-Times/next-json-ld/files/8621750/log.json.txt)

juanSanchezAlcala commented:
I've made an update on report files and code

GlynnPhillips commented on types/breadcrumb.js:
```diff
@@ -0,0 +1,108 @@
+	});
+}
+function getBreadcrumbItems (content){
+	let items = getBreadcrumbItemsFromAnnotation(content.annotations);
```
**Suggestion:** I would prefer this to be reworked into a `const`. The problem I have with a `let` is that I now need to keep a mental model of what this could change into at any stage. Using `const` means I can better understand what each value will be at any stage in its lifecycle

GlynnPhillips commented on types/breadcrumb.js:
```diff
@@ -0,0 +1,108 @@
+	return annotations.map(annotation => {
+		annotation.predicateName = annotation.predicate.split('/').pop();
+		return annotation;
+	}).filter(annotation => Object.keys(predicateWeights).includes(annotation.predicateName)).sort(comparePredicates)
```
**Suggestion:** There is a lot happening here and it is therefore difficult to understand each part. I think we could optimize this better for readability and future maintenance. Maybe something like this would be better



```js
const highlLevelAnnotations = annotations
    .filter(annotation => predicateWeights[annotation.predicateName]
);
const orderedAnnotations = highlLevelAnnotations.sort(orderByWeighting);

```

GlynnPhillips commented on types/breadcrumb.js:
```diff
@@ -0,0 +1,108 @@
+
+const maxTagNumber = 3;
+
+function countSlashes (str){
```
**Suggestion:** To prevent confusion please use verbose naming `string`

GlynnPhillips commented on types/breadcrumb.js:
```diff
@@ -0,0 +1,108 @@
+
+function countSlashes (str){
+	if(!str) return 0;
+	return (str.match(/\//g) || []).length;
```
**Suggestion:** I know this is a simple one but I always try to avoid regex where possible. Maybe using split would be nicer to read it would also mean you don't need the `||` as split will always return us an array.


```js
return str.split('/').length - 1;
```

juanSanchezAlcala commented on types/breadcrumb.js:
```diff
@@ -0,0 +1,108 @@
+function getBreadcrumbItems (content){
+	let items = getBreadcrumbItemsFromAnnotation(content.annotations);
+	let lastItem = getBreadcrumbItemFromDisplayConcept(content.displayConcept,items.length + 1);
+	if(lastItem){
+		let found = false;
+		if(found = items.find(item => item.name === lastItem.name ))
+			items.splice(items.indexOf(found),1);
+		else
+			items.pop();
+		items.push(lastItem);
+		items = repositioning(items);
+	}
+	return items;
```
> **Suggestion:** I found this section very difficult to follow and I had to run it to understand what is happening. I think this could be simplified to prioritise readability.
> 
> It would also be nicer to understand if there was less mutation happening to the array and we were explicit about when we are creating a new array with different values.

Here what im doing is check if inside "items" exists "lastItem" and if it exists delete because we need to set it into the last position and have a maximum of 3 items . Maybe sometimes we don't have "lastItem" because it comes from displayConcept property and maybe this property do not exists always so we pick the three elements from annotations . Due to this we need to recalculate the position of each element because we need to set position right. 

I'll try to find a way easier to do it but maybe i do not find a good way to do with a "items" as a const. Let me review and I'll notice you

juanSanchezAlcala commented:
I think i've almost solved all the issues you appoint in the code . I refactored the code to be more understandable and when it posible use const , also i have added more tests. Can you review it again ?

GlynnPhillips commented on types/breadcrumb.js:
```diff
@@ -0,0 +1,97 @@
+function getBreadcrumbItems (content){
+	const items = getItems(content);
+	return items.map((annotation,index) => {
+		return getBreadcrumbItem(index + 1,annotation.prefLabel,annotation.url);
```
**Suggestion:** It would be great go set `index + 1` as a named variable so its clear what this is being used for without needing to look at the function that receives it.

```js
const position = index + 1;
```

GlynnPhillips commented on types/breadcrumb.js:
```diff
@@ -0,0 +1,97 @@
+	return highlLevelAnnotations.sort(orderByPredicateName).slice(0,maxTagNumber);
+}
+
+function getItemsFromAnnotation (annotations){
+	if(annotations){
+		return getHierarchyAnnotations(annotations);
+	}
+	return [];
+}
```
**Suggestion:** I feel like this function is an unnecessary abstraction and could be be put into the `getHierarchyAnnotations` function

```js
function getHierarchyAnnotations (annotations) {
	if (!annotations || (annotations && !annotations.length)) {
             return [];
        }
	const highlLevelAnnotations = annotations.map(annotation => {
		annotation.predicateName = annotation.predicate.split('/').pop();
		return annotation;
	}).filter(annotation => predicateWeights[annotation.predicateName]);

	return highlLevelAnnotations.sort(orderByPredicateName).slice(0,maxTagNumber);
}
```

GlynnPhillips reviewed (changes requested):
Couple of small suggestions

juanSanchezAlcala commented:
Yes you are right , I've just made the changes.
